File: Java/PdfViewerAPI_R2/java/com/pdf_tools/pdfviewer/Model/PdfTextFragment.java
package com.pdf_tools.pdfviewer.Model;

import java.util.Arrays;
import java.util.List;

import com.pdf_tools.pdfviewer.converter.geom.Rectangle;

public class PdfTextFragment extends NativeObject
{

    PdfTextFragment(String text, int pageNo, double dX, double dY, double dWidth, double dHeight, Double[] pOffsets)
    {
        rectOnUnrotatedPage = new Rectangle.Double(dX, dY, dWidth, dHeight);
        this.text = text;
        this.pageNo = pageNo;
        glyphPositions = Arrays.asList(pOffsets);
    }

    private boolean finalized = false;

    @Override
    public void finalize()
    {
        if (!finalized)
            releaseTextFragment(this);
        finalized = true;
    }

    public Rectangle.Double getRectOnUnrotatedPage()
    {
        return rectOnUnrotatedPage;
    }

    public Rectangle.Double getRectOnUnrotatedPage(int start, int end)
    {
        end = Math.min(end, glyphPositions.size() - 1);
        Rectangle.Double r = (Rectangle.Double) rectOnUnrotatedPage.clone();
        r.x = glyphPositions.get(start);
        r.width = glyphPositions.get(end) - glyphPositions.get(start);
        return r;
    }

    public String getText()
    {
        return text;
    }

    public int getPageNo()
    {
        return pageNo;
    }

    public String toString()
    {
        return "Text=" + text + ", rectOnUnrotatedPage=" + rectOnUnrotatedPage.toString();
    }

    public int getIndexOfClosestGlyph(double x)
    {
        int start = 0;
        int end = glyphPositions.size() - 1;

        while (true)
        {
            if (end - start < 2)
            {
                // only end and start are the closest indices
                if (Math.abs(glyphPositions.get(end) - x) < Math.abs(glyphPositions.get(start) - x))
                    return end;
                return start;
            }
            int needle = (end - start) / 2 + start;
            switch (Double.compare(x, glyphPositions.get(needle)))
            {
            case 1:
                start = needle;
                break;
            case -1:
                end = needle;
                break;
            case 0:
                return needle;
            }
        }
    }

    private List<Double> glyphPositions;
    private Rectangle.Double rectOnUnrotatedPage;
    private String text;
    private int pageNo;

}

File: CS.NET/PdfViewerCSharpAPI_R2/Utilities/PdfTextFragment.cs
// -----------------------------------------------------------------------
// <copyright file="PdfTextFragment.cs" company="">
// TODO: Update copyright text.
// </copyright>
// -----------------------------------------------------------------------

namespace PdfTools.PdfViewerCSharpAPI.Utilities
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using PdfTools.PdfViewerCSharpAPI.DocumentManagement;

    /// <summary>
    /// TODO: Update summary.
    /// </summary>
    public class PdfTextFragment
    {
        public PdfTextFragment(PdfDocument.NativeTextFragment frag, int pageNo, PdfTextFragment previousTextFragment)
        {
            rectOnUnrotatedPage = new PdfSourceRect(frag.m_dX, frag.m_dY, frag.m_dWidth, frag.m_dHeight);
            text = frag.m_szTextA;
            this.pageNo = pageNo;
            _lastOnLine = this;

            glyphPositions = new double[frag.m_nGlyphPositionSize];
            Array.Copy(frag.m_pdGlyphPosition, glyphPositions, frag.m_nGlyphPositionSize);

            //if firstOnLine is entirely above this
            if (previousTextFragment==null || previousTextFragment.FirstOnLine.RectOnUnrotatedPage.dY > this.RectOnUnrotatedPage.dBottom)
            {
                _firstOnLine = this;
            }
            else
            {
                _firstOnLine = previousTextFragment.FirstOnLine;
                previousTextFragment._lastOnLine = this;
            }
        }

        public PdfSourceRect RectOnUnrotatedPage
        {
            get
            {
                return rectOnUnrotatedPage;
            }
        }

        public PdfSourceRect GetRectOnUnrotatedPage(int start, int end)
        {
            end = Math.Min(end, glyphPositions.Length - 1);
            PdfSourceRect rect = rectOnUnrotatedPage.Clone();
            rect.dX = glyphPositions[start];
            rect.dWidth = glyphPositions[end] - glyphPositions[start];
            return rect;
        }

        public int GetIndexOfClosestGlyph(double x)
        {
            int start = 0;
            int end = glyphPositions.Length - 1;
            
            while (true)
            {
                if (end - start < 2)
                {
                    //only end and start are the closest indices
                    if(Math.Abs(glyphPositions[end] - x) < Math.Abs(glyphPositions[start] - x))
                        return end;
                    return start;
                }
                int needle = (end - start) / 2 + start;
                switch(x.CompareTo(glyphPositions[needle])){
                    case 1:
                        start = needle;
                        break;
                    case -1:
                        end = needle;
                        break;
                    case 0:
                        return needle;
                }
            }

        }

        public string Text
        {
            get
            {
                return text;
            }
        }
        public int PageNo
        {
            get
            {
                return pageNo;
            }
        }

        PdfTextFragment _firstOnLine;
        public PdfTextFragment FirstOnLine
        {
            get
            {
                return _firstOnLine;
            }
        }
        PdfTextFragment _lastOnLine;
        public PdfTextFragment LastOnLine
        {
            get
            {
                if(_lastOnLine != this)
                    _lastOnLine = _lastOnLine.LastOnLine;
                return _lastOnLine;
            }
        }


        public override string ToString()
        {
            return String.Format("Text={0}, rectOnPage={1}", text, rectOnUnrotatedPage.ToString());
        }

        private PdfSourceRect rectOnUnrotatedPage; //These are source coordinates with origin in bottom left!
        private string text;
        private int pageNo;
        private double[] glyphPositions;
    }
}
